### PR TITLE
fix(binder): insert returning column layout 

### PR DESCRIPTION
**Pull request by broccoliSpicy**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

 fix issue #9012 

try to fix issue #9012 by faking a bind context that aligns with the inserting values schema then bind returning list, 
so that every InputRef in returning_list can have a correct index referring to the insert values.

disabled column pruning for insertion->projection, diffident/fearing about this and I could really use some discussion 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/binder/insert.rs [hunk 1]
```diff
@@ -143,6 +140,17 @@ impl Binder {
...
 
+        let returning_list;
+        let fields;
+        // since the input value's column layout may differ from the target table's column layout,
+        // we mock a column layout that aligns with the insert values when binding return_lists.
+        {
+            let original_ctx = self.context.clone();
+            self.context.arrange_column_layout(&col_indices_to_insert);
+            (returning_list, fields) = self.bind_returning_list(returning_items)?;
+            self.context = original_ctx;
```
Comment (BugenZhao) [inline src/frontend/src/binder/insert.rs:151]: I'm not that familiar with this, but is `push_context` applicable here?

File: src/frontend/src/optimizer/plan_node/logical_project.rs [hunk 2]
```diff
@@ -148,6 +148,14 @@ impl fmt::Display for LogicalProject {
 
 impl ColPrunable for LogicalProject {
     fn prune_col(&self, required_cols: &[usize], ctx: &mut ColumnPruningContext) -> PlanRef {
+        // for example, in query:
+        // `insert into t values(v1, v2) returning v1;`
+        // column pruning may mess up the insert values.
+        // so column pruning is disabled in insert->project
+        if self.input().node_type() == PlanNodeType::LogicalInsert {
+            return self.clone().into();
+        }
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_project.rs:157]: I think we can not do it, especially in the `Project::prune_col`. Instead, we need to change the `Insert::prune_col`. 
e.g. for the `insert into t values(v1, v2) returning v2*2;` we will have the following plan 
```SQL
Project(input_ref(1)*2)
    Insert(v1, v2)
         Values(v1, v2)
```
and the `Project::prune_col` will call its child, `Insert::prune_col` with the required column [1]. the Insert::prune_col need to return a plan only with the column 0. the plan is expected to be 
```SQL
Project(input_ref(0)*2)
    Project(input_ref(1))
        Insert(v1, v2)
            Values(v1, v2)
```

You can refer to https://github.com/risingwavelabs/risingwave/blob/main/src/frontend/src/optimizer/plan_node/logical_project_set.rs#L221-L230

Comment (broccoliSpicy) [reply src/frontend/src/optimizer/plan_node/logical_project.rs]: indeed, modified.

**Comment by broccoliSpicy** [comment]
the current CI failure is under discussion in #9539

**Comment by Eurekaaw** [comment]
Hello @broccoliSpicy , do you have any updates on this issue?

**Comment by fuyufjh** [comment]
Hi, it's been a long time. Any updates?

**Pull request closed without merging**